compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Prefixing.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE GADTs #-}
+
+-- This module handles prefixing of identifiers generated by the SubIDEs, so that they are unique to the client, but are returned in the same form to the servers.
+module DA.Cli.Damlc.Command.MultiIde.Prefixing (
+  addProgressTokenPrefixToServerMessage,
+  addLspPrefixToServerMessage,
+  removeLspPrefix,
+  removeWorkDoneProgressCancelTokenPrefix,
+) where
+
+import Control.Concurrent.MVar
+import Control.Lens
+import Data.Foldable (find)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import qualified Data.Text as T
+import qualified Language.LSP.Types as LSP
+import qualified Language.LSP.Types.Lens as LSP
+import DA.Cli.Damlc.Command.MultiIde.Types
+
+-- ProgressToken Prefixing
+
+-- We need to ensure all ProgressTokens from different subIDEs are unique to the client, so we prefix them.
+-- We do not need to undo this transformation, as we'll store it and backwards lookup later.
+addProgressTokenPrefix :: T.Text -> LSP.ProgressToken -> LSP.ProgressToken
+addProgressTokenPrefix prefix (LSP.ProgressNumericToken t) = LSP.ProgressTextToken $ prefix <> "-" <> T.pack (show t)
+addProgressTokenPrefix prefix (LSP.ProgressTextToken t) = LSP.ProgressTextToken $ prefix <> "-" <> t
+
+-- Added to WindowWorkDoneProgressCreate and Progress
+-- If its create, add to the var, if its progress, read from the var, fall back to current
+addProgressTokenPrefixToServerMessage :: ServerCreatedProgressTokensVar -> FilePath -> T.Text -> LSP.FromServerMessage -> IO LSP.FromServerMessage
+addProgressTokenPrefixToServerMessage tokensVar home prefix (LSP.FromServerMess LSP.SWindowWorkDoneProgressCreate req) = do
+  let unPrefixedToken = req ^. LSP.params . LSP.token
+      prefixedToken = addProgressTokenPrefix prefix unPrefixedToken
+  modifyMVar_ tokensVar (pure . Map.insert (unPrefixedToken, home) prefixedToken)
+  pure $ LSP.FromServerMess LSP.SWindowWorkDoneProgressCreate $ req & LSP.params . LSP.token .~ prefixedToken
+addProgressTokenPrefixToServerMessage tokensVar home _ (LSP.FromServerMess LSP.SProgress notif) = do
+  let unPrefixedToken = notif ^. LSP.params . LSP.token
+  prefixedToken <- fromMaybe unPrefixedToken . Map.lookup (unPrefixedToken, home) <$> readMVar tokensVar
+  case notif ^. LSP.params . LSP.value of
+    LSP.End _ -> modifyMVar_ tokensVar (pure . Map.delete (unPrefixedToken, home))
+    _ -> pure ()
+  pure $ LSP.FromServerMess LSP.SProgress $ notif & LSP.params . LSP.token .~ prefixedToken
+addProgressTokenPrefixToServerMessage _ _ _ msg = pure msg
+
+findKeyByValue :: Eq v => v -> Map.Map k v -> Maybe k
+findKeyByValue val = fmap fst . find ((==val) . snd) . Map.toList
+
+-- Remove from WindowWorkDoneProgressCancel
+removeWorkDoneProgressCancelTokenPrefix
```

Comment:
Sam note: Detail why the prefixing logic is need in one direction (subIDE to client) but isn't necessary in the other direction